### ISSUE TITLE
brig cat: big files are only outputted partially

### ISSUE THREAD

**Issue by sahib** [sahib/brig#70]
**Describe the bug**

I added a big video (2G) via ``brig stage /file.mkv`` file and ``brig cat /file.mkv`` only outputted first 37 MB. Sometimes the size is also bigger. The syslog shows this line indicating a badly encoded stream or bad decoding:

```
28.12.2020/16:20:48 ⚠ server/fs_handler.go:172: IO failed for path /file.mkv on 127.0.0.1:46267: nonce size mismatch; should: 12 -have: 4
```

**To Reproduce**

See above.

**Expected behavior**

The full file to be outputted.

**Additional context**

Both IPFS and brig repositories are in /tmp.

**Comment by evgmik** [comment]
Confirming. Mine crashed at around 1.1 GB mark. With same complain  about `nonce size mismatch; should: 12 - have: 4`

An interesting observation:
after I did `ipfs cat backendhash` , the `brig cat` went trough just fine

**Comment by sahib** [comment]
Might have found something related to that while working on the encryption scheme:

I tried to use ``io.MultiReader`` to prepend the header buffer (that we use to guess the mime) to the stream. That did not work and broke one test since only part of the stream could be read back. I then switched from ``io.MultiReader`` to our own ``prefixReader`` implementation. The difference here is:

* ``io.MultiReader``: ``Read()`` will only return results from one of its readers. So when it's exhausted, it will read the rest of it until io.EOF, return and on the next ``Read()`` the next reader will be tried.
* ``prefixReader``: Is a bit smarter and empties the buffer first and if there's still place in the read buffer it will read from the underlying stream. 

This hints to the compress writer not correctly cases when the underlying reader does not fill the buffer completely. Typical mistake here would be to check the buffer length, but not the length that ``Read()`` returned. Maybe also the encryption writer is affected, but I commented that one out for testing and showed primarily with compression. Without compression or encryption it worked nicely (at least in the test, did not check with the 2G file).

Next data point: The content hash for the dummy data was always the same, but the backend hash differed when changing the header buffer (512 worked, 1024 did not) while still using ``io.MultiReader``. This is probably because 512 is a multiple of the buffer size inside of the compression writer, but 1024 is not.

> An interesting observation:
> after I did ipfs cat backendhash , the brig cat went trough just fine

Yes, that's indeed interesting. Hints towards the same reason.

**Comment by sahib** [comment]
I thought I fixed it with ae56b750, but apparently there's another bug hiding. Works now for some basic files and tests (no matter if I use ``io.MultiReader`` or not), but the 2G file still fails. Should probably try around with that a bit.

**Comment by sahib** [comment]
While searching for this bug I found another bug, that sometimes could happen for large read buffers: 4ab362aca14dd4d867fd331bc85902a266e320ee - but also that one does not fix the issue.

I now verified the following things:

* The bug also happens for files with 256M, probably less too but less often. With 256M it happens more often than not for me.
* The file is correctly encoded and uploaded to IPFS. I can manually ``ipfs cat`` it and pipe it through our decoding logic.
* Introducing debug logs or small sleeps in [here](https://github.com/sahib/brig/blob/develop/backend/httpipfs/io.go#L37) "fix" the bug.
* My attempt to build with race detection was futile since it slows down execution a little, making things work. :man_facepalming: 

I suspect this is something that happened with the upgrade to IPFS 0.7.0. So my plan is to check if older versions of IPFS do not show this behavior to get a clue what could be wrong. It's definitely a timing related issue, most likely in ``backend/httpipfs/io.go``
